v2/sbomgcr/sbom_payload.go

```diff
@@ -0,0 +1,152 @@
+package sbomgcr
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/CycloneDX/cyclonedx-go"
+	"github.com/mitchellh/mapstructure"
+)
+
+// structs for unmarshalling
+
+// Envelope is the envelope of a signed payload of the SBOM
+type Envelope struct {
+	PayloadType string      `json:"payloadType"`
+	Payload     string      `json:"payload"`
+	Signatures  []Signature `json:"signatures"`
+}
+
+// Signature is the signature found in the envelope of the SBOM
+type Signature struct {
+	KeyID string `json:"keyid"`
+	Sig   string `json:"sig"`
+}
+
+// PayloadPredicateType represents the type of Predicate in the payload
+type PayloadPredicateType struct {
+	PredicateType string `json:"predicateType"`
+}
+
+// SBOMPredicate is an interface for the SBOM Predicates voucher supports
+type SBOMPredicate interface {
+	GetPredicateType() string
+	ToBOM(sbom *cyclonedx.BOM) error
+}
+
+// CustomPredicate represents a custom sbom predicate
+type CustomPredicate struct {
+	Type          string `json:"_type"`
+	PredicateType string `json:"predicateType"`
+	Subject       []struct {
+		Name   string `json:"name"`
+		Digest struct {
+			Sha256 string `json:"sha256"`
+		} `json:"digest"`
+	} `json:"subject"`
+	Predicate struct {
+		Data      string    `json:"Data"`
+		Timestamp time.Time `json:"Timestamp"`
+	} `json:"predicate"`
+}
+
+func (cp CustomPredicate) GetPredicateType() string {
+	return cp.PredicateType
+}
+
+func (cp CustomPredicate) ToBOM(sbom *cyclonedx.BOM) error {
+	return json.Unmarshal([]byte(cp.Predicate.Data), &sbom)
+}
+
+// CycloneDXPredicate represents a cyclonedx sbom predicate
+type CycloneDXPredicate struct {
+	Type          string `json:"_type"`
+	PredicateType string `json:"predicateType"`
+	Subject       []struct {
+		Name   string `json:"name"`
+		Digest struct {
+			Sha256 string `json:"sha256"`
+		} `json:"digest"`
+	} `json:"subject"`
+	Predicate struct {
+		Data map[string]interface{} `json:"Data"`
+	} `json:"predicate"`
+}
+
+func (cyclonedxp *CycloneDXPredicate) GetPredicateType() string {
+	return cyclonedxp.PredicateType
+}
+
+func (cyclonedxp *CycloneDXPredicate) ToBOM(sbom *cyclonedx.BOM) error {
+	return mapstructure.Decode(cyclonedxp.Predicate.Data, &sbom)
+}
+
+func getEnvelopeFromReader(reader io.ReadCloser) (Envelope, error) {
```

Comment:
```suggestion
func getEnvelopeFromReader(reader io.Reader) (Envelope, error) {
```

Idea: if you don't intend to interact with `.Close()`, just accept an `io.Reader`.

From this signature, I'd expect the function to `defer reader.Close()` for me.